R/tm_data_table.R

```diff
@@ -305,36 +323,65 @@ srv_data_table <- function(id,
                            if_distinct,
                            dt_args,
                            dt_options,
-                           server_rendering) {
+                           server_rendering,
+                           decorators) {
   moduleServer(id, function(input, output, session) {
     iv <- shinyvalidate::InputValidator$new()
     iv$add_rule("variables", shinyvalidate::sv_required("Please select valid variable names"))
     iv$add_rule("variables", shinyvalidate::sv_in_set(
       set = names(isolate(data())[[dataname]]), message_fmt = "Not all selected variables exist in the data"
     ))
     iv$enable()
+    # teal::validate_inputs(iv)
 
-    output$data_table <- DT::renderDataTable(server = server_rendering, {
-      teal::validate_inputs(iv)
-
+    data_table_data <- reactive({
       df <- data()[[dataname]]
-      variables <- input$variables
 
       teal::validate_has_data(df, min_nrow = 1L, msg = paste("data", dataname, "is empty"))
 
-      dataframe_selected <- if (if_distinct()) {
-        dplyr::count(df, dplyr::across(dplyr::all_of(variables)))
-      } else {
-        df[variables]
-      }
+      teal.code::eval_code(
+        data(),
+        substitute(
+          expr = {
+            variables <- vars
+            dataframe_selected <- if (if_distinct) {
+              dplyr::count(dataname, dplyr::across(dplyr::all_of(variables)))
+            } else {
+              dataname[variables]
+            }
+            dt_args <- args
+            dt_args$options <- dt_options
+            if (!is.null(dt_rows)) {
+              dt_args$options$pageLength <- dt_rows
+            }
+            dt_args$data <- dataframe_selected
+            table <- do.call(DT::datatable, dt_args)
+          },
+          env = list(
+            dataname = as.name(dataname),
+            if_distinct = if_distinct(),
+            vars = input$variables,
+            args = dt_args,
+            dt_options = dt_options,
+            dt_rows = input$dt_rows
+          )
+        )
+      )
+    })
 
-      dt_args$options <- dt_options
-      if (!is.null(input$dt_rows)) {
-        dt_args$options$pageLength <- input$dt_rows
-      }
-      dt_args$data <- dataframe_selected
+    decorated_data_table_data <-
+      srv_teal_transform_data("decorate", data = data_table_data, transformators = decorators)
 
-      do.call(DT::datatable, dt_args)
+    output$data_table <- DT::renderDataTable(server = server_rendering, {
+      req(data_table_data())
+      # no table is displayed
+      decorated_data_table_data()[["table"]]
     })
+
+    teal.widgets::verbatim_popup_srv(
+      id = "rcode",
+      verbatim_content = reactive(teal.code::get_code(req(decorated_data_table_data()))),
+      title = "R Code for PCA"
+    )
```

Comment:
"Show R Code is not working for me, can you confirm it?

Do we need to have the button? it wasn't there before

Reply:
@averissimo the Show R Code was in the wrong UI place and there was a namespace collision. Just moved it https://github.com/insightsengineering/teal.modules.general/pull/799/commits/d769821f66c71bb33365a78aec1884fff5ee5c4f

But the code looks super weird. I think we should put if statements around `eval_code` and not inside `eval_code`

<img width="340" alt="image" src="https://github.com/user-attachments/assets/7a4659a2-222f-40f3-bb95-2db0d83c2253">

Reply:
Check out all those weird `if(FALSE)` statements or `if(!is.null(NULL)`.
But I'm not sure it's worth it.

The decorator code, for example would need to look like this

```r
      reactive({
        req(data())
        
        if (input$style == "Striped") {
          within(data(), 
           {
             
               table <-
                 DT::formatStyle(
                   table,
                   columns = attr(table$x, "colnames")[-1],
                   target = 'row',
                   backgroundColor = '#f9f9f9'
                 )
             
           }
          )
        }  else if (input$style == "Hover") {
          within(data(), 
                 {
                  table <-
                    DT::formatStyle(
                      table,
                      columns = attr(table$x, "colnames")[-1],
                      target = 'row',
                      backgroundColor = '#f0f0f0'
                    )
                 }
          )
        } else {
          data()
        }
          
      })
```

instead of

```r
      reactive({
        req(data(), input$style)
        within(data(), 
               {
                 if (style == "Striped") {
                   table <-
                     DT::formatStyle(
                       table,
                       columns = attr(table$x, "colnames")[-1],
                       target = 'row',
                       backgroundColor = '#f9f9f9'
                     )
                 } else if (style == "Hover") {
                   table <-
                     DT::formatStyle(
                       table,
                       columns = attr(table$x, "colnames")[-1],
                       target = 'row',
                       backgroundColor = '#f0f0f0'
                     )
                 }
               }, 
               style = input$style
        )
      })
```

Reply:
I don't think we should add new functionality to the module on the PR.

If you are feeling conflicted open a new issue.